### PR TITLE
[webgpu] Add modularized nonMaxSuppression kernel

### PR DESCRIPTION
**Pull request by axinging**
FIX https://github.com/tensorflow/tfjs/issues/2361

nonMaxSuppression is implemented by CPU. 
When running in WebGL backend, WebGL supports readSync. If passing any tensors reside in GPU to it, nonMaxSuppression can automatically read back GPU data to CPU.

When running in WebGPU backend, WebGPU doesn't support readSync. If passing any tensors reside in GPU to it directly, nonMaxSuppression may complains "WebGPU readSync is only available for CPU-resident tensors".  So before this, call await first to gurantee the tensors are in CPU.

The related issue for this: https://github.com/tensorflow/tfjs/issues/2976.

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/2963)

### CODE REVIEW

**Comment by axinging** [comment]
@annxingyuan PTAL
It seems build fail on the Trigger.

**Comment by axinging** [comment]
@annxingyuan，the build error is fixed. PTAL.

**Comment by annxingyuan** [comment]
@axinging would you mind merging in master first before we take a look? we've just reorganized everything - I should have given you a heads up. hopefully the merge is easy!

**Comment by axinging** [comment]
@annxingyuan , reabse is done, thanks. PTAL.

**Comment by axinging** [comment]
@annxingyuan, done,PTAL

File: tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV3.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =============================================================================
+ */
+
+import {backend_util, KernelConfig, NonMaxSuppressionV3, NonMaxSuppressionV3Attrs, NonMaxSuppressionV3Inputs} from '@tensorflow/tfjs-core';
+import {WebGPUBackend} from '../backend_webgpu';
+export type TypedArray = Float32Array|Int32Array|Uint8Array;
```
Comment (annxingyuan) [inline tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV3.ts:20]: Why define `TypedArray` here? Can you import `TypedArray` from tfjs-core instead?

File: tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV3.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    console.warn(
+        'tf.nonMaxSuppression() in webgpu locks the UI thread. ' +
+        'Call tf.nonMaxSuppressionAsync() instead');
+
+    const {boxes, scores} = inputs as NonMaxSuppressionV3Inputs;
+    const {maxOutputSize, iouThreshold, scoreThreshold} =
+        attrs as unknown as NonMaxSuppressionV3Attrs;
+
+    const gpuBackend = backend as WebGPUBackend;
+    const boxesVals = gpuBackend.readSync(boxes.dataId) as TypedArray;
```
Comment (annxingyuan) [inline tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV3.ts:34]: This only works if boxes and scores are on the CPU right? Can we add a check that ensures they are on the CPU and throw an error if they're not?

File: tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV5.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        'tf.nonMaxSuppression() in webgpu locks the UI thread. ' +
+        'Call tf.nonMaxSuppressionAsync() instead');
+
+    const {boxes, scores} = inputs as NonMaxSuppressionV5Inputs;
+    const {maxOutputSize, iouThreshold, scoreThreshold, softNmsSigma} =
+        attrs as unknown as NonMaxSuppressionV5Attrs;
+
+    const gpuBackend = backend as WebGPUBackend;
+
+    const boxesVals = gpuBackend.readSync(boxes.dataId) as TypedArray;
```
Comment (annxingyuan) [inline tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV5.ts:36]: Same comment as above - should we throw an error if the data is not on the CPU?

**Comment by annxingyuan** [comment]
thanks @axinging and really sorry about the delay on this - I've added my teammate Na to the PR - she has more context about the non max suppression kernel!

File: tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV3.ts (hunk 2 above)
Comment (axinging) [reply tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV3.ts:34]: WebGPU::readSync will through a error when the box or score are in not the GPU:
WebGPU readSync is only available for CPU-resident tensors.

File: tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV3.ts (hunk 1 above)
Comment (axinging) [reply tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV3.ts]: Done here: https://github.com/tensorflow/tfjs/pull/2963/commits/756630081206adef8721d2b4055e6e0f113b7e59

File: tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV5.ts (hunk 3 above)
Comment (axinging) [reply tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV5.ts:36]: WebGPU backend has handle this.

**Comment by axinging** [comment]
> Hi @axinging , in general LGTM. Thank you! It's very neat! I left a few minor comments. Main thing required is to import names from kernel_name.ts instead of using string values. According to @tafsiri, by "exporting all kernel names as symbols that can be imported by ops. This provides a way to map from ops to all kernels that are used."
> 
> > _[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/2963#-:-M6cK4Z44E_Mu-RpE2TX:b7yd209)_ status: 0 of 1 approvals obtained (waiting on @axinging and @lina128)
> 
> _[tfjs-core/src/ops/NonMaxSuppression.ts, line 38 at r3](https://reviewable.io/reviews/tensorflow/tfjs/2963#-M6cGh_V0aJYyFOllrqe:-M6cGh_V0aJYyFOllrqf:b-f522oe) ([raw file](https://github.com/tensorflow/tfjs/blob/f968aecd9b63272e0076f8ec3fb974e90460a7f9/tfjs-core/src/ops/NonMaxSuppression.ts#L38)):_
> 
> > ```
> >   scoreThreshold = inputs.scoreThreshold;
> > 
> >   const attrs = {maxOutputSize, iouThreshold, scoreThreshold};
> > ```
> 
> Can you add a NonMaxSuppressionV3Attrs interface in kernel_name.ts? Add a NonMaxSuppressionV3Inputs type too.
> 
> _[tfjs-core/src/ops/NonMaxSuppression.ts, line 42 at r3](https://reviewable.io/reviews/tensorflow/tfjs/2963#-M6cGuky3ZV5DgstJCRw:-M6cGuky3ZV5DgstJCRx:b-augcol) ([raw file](https://github.com/tensorflow/tfjs/blob/f968aecd9b63272e0076f8ec3fb974e90460a7f9/tfjs-core/src/ops/NonMaxSuppression.ts#L42)):_
> 
> > ```
> >       b => b.nonMaxSuppression(
> >           $boxes, $scores, maxOutputSize, iouThreshold, scoreThreshold),
> >       {boxes: $boxes, scores: $scores}, null /* grad */, 'NonMaxSuppressionV3',
> > ```
> 
> Use NonMaxsuppressionV3 from kernel_name.ts, not the string value.

Thanks, Done here: https://github.com/tensorflow/tfjs/pull/2963/commits/97f0be7e8f99d5178e1b60b8c4e59b8371557529

**Comment by axinging** [comment]
> ```
> it('works with other operators, div'
> ```

The two are done :
1), rename NonMaxSuppression = > non_max_suppression
2), nonMaxSuppSanityCheck is used in image_op now.

About case it('works with other operators, div'):
if no "await scores.data();", this may fail on WebGPU.
This case is used to prove that, the nmx for webgpu require all input are ready in CPU(await is used to gurantee this). So for a real case, we need to remove this await. div is just an example, mul/add are all ok for this purpose.

**Comment by axinging** [comment]
@annxingyuan  @lina128  @tafsiri 
Thanks for all your commenst, PTAL.

**Comment by axinging** [comment]
@annxingyuan @tafsiri，  PTAL

**Comment by axinging** [comment]
@annxingyuan @tafsiri，updated, PTAL

**Comment by annxingyuan** [comment]
@tafsiri I think the reason NMS could interact with other operators is that NMS expects its input data to be available at the moment when it's called. In WebGPU, because everything is asynchronous, this can fail.

**Comment by axinging** [comment]
> @tafsiri I think the reason NMS could interact with other operators is that NMS expects its input data to be available at the moment when it's called. In WebGPU, because everything is asynchronous, this can fail.

For WebGPU and WebGL, below works:
```
      const a = tf.tensor1d([0, 1, -2, -4, 4, -4]);
      const b = tf.tensor1d([0.15, 0.2, 0.25, 0.5, 0.7, 1.2]);
      const scores = a.div(b);
      const maxOutputSize = 2;
      const iouThreshold = 0.5;
      const scoreThreshold = 0;
      await scores.data();
      const indices = tf.image.nonMaxSuppression(
          boxes, scores as tf.Tensor1D, maxOutputSize, iouThreshold,
          scoreThreshold);
```

For WebGL, below works (scores.data() can read GPU data into CPU, nonMaxSuppression need it in the CPU), but failed on WebGPU (scores.data() can not read back GPU data into in the CPU, nonMaxSuppression need it in the CPU):
```
      const a = tf.tensor1d([0, 1, -2, -4, 4, -4]);
      const b = tf.tensor1d([0.15, 0.2, 0.25, 0.5, 0.7, 1.2]);
      const scores = a.div(b);
      const maxOutputSize = 2;
      const iouThreshold = 0.5;
      const scoreThreshold = 0;
      scores.data(); // no await 
      const indices = tf.image.nonMaxSuppression(
          boxes, scores as tf.Tensor1D, maxOutputSize, iouThreshold,
          scoreThreshold);
```

**Comment by tafsiri** [comment]
Thx! LGTM

**Comment by tafsiri** [comment]
I would add a comment in this test for some context about this (that it is webgpu related). That would help someone not wonder about this/delete it.

**Pull request merged**